lib/propolis/src/enlightenment/hyperv/overlay.rs

```diff
@@ -0,0 +1,754 @@
+/// callout; at that point no vCPUs are active or will become active, so this
+/// should remove the last strong reference to the manager, after which it is
+/// safe to drop the pages it produced.
+///
+/// [`Lifecycle::halt`]: crate::lifecycle::Lifecycle::halt
+pub(super) struct OverlayManager {
+    inner: Mutex<ManagerInner>,
+    acc_mem: MemAccessor,
+}
+
+impl OverlayManager {
+    /// Creates a new overlay manager.
+    pub(super) fn new() -> Arc<Self> {
+        let acc_mem = MemAccessor::new_orphan();
+        Arc::new(Self { inner: Mutex::new(ManagerInner::default()), acc_mem })
+    }
+
+    /// Attaches this overlay manager to the supplied memory accessor hierarchy.
+    pub(super) fn attach(&self, parent_mem: &MemAccessor) {
+        parent_mem
+            .adopt(&self.acc_mem, Some("hyperv-overlay-manager".to_string()));
+    }
+
+    /// Adds an overlay of the supplied `kind` with the supplied `contents` over
+    /// the supplied `pfn`.
+    pub(super) fn add_overlay(
+        self: &Arc<Self>,
+        pfn: u64,
+        kind: OverlayKind,
+        contents: OverlayContents,
+    ) -> Result<OverlayPage, OverlayError> {
+        let pfn = Pfn::new(pfn)?;
+        let mut inner = self.inner.lock().unwrap();
+        inner.add_overlay(pfn, kind, contents, &self.acc_mem)?;
+        Ok(OverlayPage { manager: Arc::downgrade(self), kind, pfn })
+    }
+
+    /// Removes an overlay of the supplied `kind` from the set at the supplied
+    /// `pfn`.
+    fn remove_overlay(
+        &self,
+        pfn: Pfn,
+        kind: OverlayKind,
+    ) -> Result<(), OverlayError> {
+        self.inner.lock().unwrap().remove_overlay(pfn, kind, &self.acc_mem)
+    }
+
+    /// Moves an overlay of the supplied `kind` from `from_pfn` to `to_pfn`.
+    fn move_overlay(
+        &self,
+        from_pfn: Pfn,
+        to_pfn: Pfn,
+        kind: OverlayKind,
+    ) -> Result<(), OverlayError> {
+        self.inner.lock().unwrap().move_overlay(
+            from_pfn,
+            to_pfn,
+            kind,
+            &self.acc_mem,
+        )
+    }
+}
+
+/// Helpers for dealing with page frame numbers (guest physical page numbers, or
+/// PFNs). These help to avoid page alignment checks that would otherwise be
+/// necessary when dealing with full physical addresses.
+mod pfn {
+    use crate::{
+        common::{GuestAddr, GuestRegion, PAGE_MASK, PAGE_SHIFT, PAGE_SIZE},
+        vmm::SubMapping,
+    };
+
+    use super::{OverlayContents, OverlayError};
+
+    /// A 52-bit physical page number.
+    #[derive(Clone, Copy, Ord, PartialOrd, Eq, PartialEq)]
+    pub(super) struct Pfn(u64);
+
+    impl std::fmt::Debug for Pfn {
+        fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+            write!(f, "Pfn({:x})", self.0)
+        }
+    }
+
+    impl std::fmt::Display for Pfn {
+        fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+            write!(f, "{:x}", self.0)
```

Comment:
nitpicky: if we instead wrote
```suggestion
            fmt::LowerHex::fmt(self.0, f)
```
we would just call into the `fmt::LowerHex` implementation directly, and it would honor any additional format specifiers (e.g. `#`,  leading zeros, alignment, etc) that this was called with. 

the current implementation eats that, so if i wrote something like:
```rust
write!("{:#016}", Pfn(42));
```
and expected to get
```
0x0000000000002a
```
i would actually get
```
2a
```
instead. it _might_ be worth handling that nicer?

Reply:
Great suggestion, thanks! Done in c8ccb8b.